lib/api/core/storage/models/baseModel.js

```diff
@@ -0,0 +1,152 @@
+ * limitations under the License.
+ */
+
+'use strict';
+
+class BaseModel {
+  constructor (_source, _id) {
+    this.__id = null;
+    this.__source = {};
+
+    Reflect.defineProperty(this, '__persisted', {
+      writable: true,
+      value: false
+    });
+
+    for (const field of this.constructor.fields) {
+      //  @todo use Reflect everywhere in kuzzle
+      Reflect.defineProperty(this, field, {
+        get: () => this.__source[field],
+        set: value => {
+          this.__source[field] = value;
+        }
+      });
+    }
+
+    this._id = _id;
+    this._source = _source;
+  }
+
+  // Public methods ============================================================
+
+  async save ({ userId=null, refresh } = {}) {
+    if (! this.__persisted) {
+      const { _id, _source } = await BaseModel.indexStorage.create(
+        this.constructor.collection,
+        this._id,
+        this._source,
+        { userId, refresh });
+
+      this._id = _id;
+      this._source = _source;
+    }
+    else {
+      await BaseModel.indexStorage.update(
+        this.constructor.collection,
+        this._id,
+        this._source);
+    }
+  }
+
+  async delete ({ refresh }) {
+    if (! this.__persisted) {
+      return;
+    }
+
+    await BaseModel.indexStorage.delete(this.constructor.collection, this._id, { refresh });
+  }
+
+  serialize () {
+    return {
+      _id : this._id,
+      _source: this._source
+    };
+  }
+
+  // Getter/Setter =============================================================
+
+  get _id () {
+    return this.__id;
+  }
+
+  set _id (_id) {
+    this.__id = _id;
+  }
+
+  get _source () {
+    return this.__source;
+  }
+
+  set _source (_source) {
+    // Reflect define hidden property for hidding
+    // or symbol to keept
+    for (const key of Object.keys(_source)) {
+      if (this.constructor.fields.includes(key)) {
+        this.__source[key] = _source[key] || null;
+      }
+    }
+  }
+
+  // Static public methods =====================================================
+
+  static register (ModelClass, collection, fields) {
+    ModelClass.collection = collection;
+    ModelClass.fields = fields;
+  }
+
+  static async load (id) {
+    const result = await BaseModel.indexStorage.get(
+      this.collection,
+      id);
+
+    return this._instantiateFromDb(result);
+  }
+
+  static deleteByQuery (searchBody) {
+    return BaseModel.indexStorage.deleteByQuery(this.collection, searchBody);
+  }
+
+  static async search (searchBody, options) {
+    const resp = await BaseModel.indexStorage.search(
+      this.collection,
+      searchBody,
+      options);
+
+    return resp.hits.map(this._instantiateFromDb);
+  }
+
+  static truncate () {
+    return BaseModel.indexStorage.truncateCollection(this.collection);
+  }
+
+  // Static private methods ====================================================
+
+  static _instantiateFromDb ({ _id, _source }) {
+    const model = new this(_source, _id);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Replace this with a constructor function. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS2999)